csr_check_test.go

```diff
@@ -1992,6 +2355,177 @@ func TestAuthorizeServingRenewal(t *testing.T) {
 	}
 }
 
+func TestGetServingCert(t *testing.T) {
+	defaultPort := int32(25535)
+	defaultAddr := "127.0.0.1"
+	defaultNode := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test",
+		},
+		Status: corev1.NodeStatus{
+			Addresses: []corev1.NodeAddress{
+				{Type: corev1.NodeInternalIP, Address: defaultAddr},
+			},
+			DaemonEndpoints: corev1.NodeDaemonEndpoints{
+				KubeletEndpoint: corev1.DaemonEndpoint{
+					Port: defaultPort,
+				},
+			},
+		},
+	}
+
+	wrongAddr := defaultNode.DeepCopy()
+	wrongAddr.Status.DaemonEndpoints.KubeletEndpoint.Port = int32(25544)
+
+	uninitialized := defaultNode.DeepCopy()
+	uninitialized.Status = corev1.NodeStatus{}
+
+	tests := []struct {
+		name      string
+		nodeName  string
+		node      *corev1.Node
+		rootCerts []*x509.Certificate
+		wantErr   string
+		serverErr error
+	}{
+		{
+			name:      "all good",
+			nodeName:  "test",
+			node:      defaultNode,
+			rootCerts: []*x509.Certificate{parseCert(t, rootCertGood)},
+		},
+		{
+			name:      "unknown certificate",
+			nodeName:  "test",
+			node:      defaultNode,
+			rootCerts: []*x509.Certificate{parseCert(t, differentCert)},
+			wantErr:   "x509: certificate signed by unknown authority",
+		},
+		{
+			name:      "node not found",
+			nodeName:  "test",
+			rootCerts: []*x509.Certificate{parseCert(t, rootCertGood)},
+			wantErr:   "nodes \"test\" not found",
+			serverErr: fmt.Errorf("nodes \"test\" not found"),
+		},
+		{
+			name:      "wrong address",
+			nodeName:  "test",
+			node:      wrongAddr,
+			rootCerts: []*x509.Certificate{parseCert(t, rootCertGood)},
+			wantErr:   "dial tcp 127.0.0.1:25544: connect: connection refused",
+		},
+		{
+			name:     "no pool provided",
+			nodeName: "test",
+			node:     wrongAddr,
+			wantErr:  "no CA found: will not retrieve serving cert",
+		},
+		{
+			name:      "node with no addr",
+			nodeName:  "test",
+			node:      uninitialized,
+			rootCerts: []*x509.Certificate{parseCert(t, rootCertGood)},
+			wantErr:   "node test has no internal addresses",
+		},
+	}
+
+	server := fakeResponder(t, fmt.Sprintf("%s:%v", defaultAddr, defaultPort), serverCertGood, serverKeyGood)
+	defer server.Close()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var certPool *x509.CertPool
+			if len(tt.rootCerts) > 0 {
+				certPool = x509.NewCertPool()
+				for _, cert := range tt.rootCerts {
+					certPool.AddCert(cert)
+				}
+			}
+
+			nodes := &testNode{t: t, node: tt.node, err: tt.serverErr}
+
+			go respond(server)
```

Comment:
is there any chance we would deadlock here inside of `respond`? (i saw an `accept` call in that function)

Reply:
Accept returns an error, once the socket is closed. I ignore it, as I only needed a consumer to return some response with peer certificates (if it is needed, and I call it only when it is needed)